python/ray/data/_internal/util.py

```diff
@@ -37,31 +44,46 @@ def _lazy_import_pyarrow_dataset() -> LazyModule:
 
 def _check_pyarrow_version():
     global _VERSION_VALIDATED
+
     if not _VERSION_VALIDATED:
-        import pkg_resources
+        if os.environ.get(RAY_DISABLE_PYARROW_VERSION_CHECK, "0") == "1":
+            _VERSION_VALIDATED = True
+            return
 
         try:
-            version_info = pkg_resources.require("pyarrow")
-            version_str = version_info[0].version
-            version = tuple(int(n) for n in version_str.split(".") if "dev" not in n)
-            if version < MIN_PYARROW_VERSION:
-                raise ImportError(
-                    "Datasets requires pyarrow >= "
-                    f"{'.'.join(str(n) for n in MIN_PYARROW_VERSION)}, "
-                    f"but {version_str} is installed. Upgrade with "
-                    "`pip install -U pyarrow`."
-                )
-        except pkg_resources.DistributionNotFound:
+            import pyarrow
+        except ModuleNotFoundError:
+            # pyarrow not installed, short-circuit.
+            return
+
+        import pkg_resources
+
+        if not hasattr(pyarrow, "__version__"):
             logger.warning(
-                "You are using the 'pyarrow' module, but "
-                "the exact version is unknown (possibly carried as "
-                "an internal component by another module). Please "
-                "make sure you are using pyarrow >= "
-                f"{'.'.join(str(n) for n in MIN_PYARROW_VERSION)} "
-                "to ensure compatibility with Ray Datasets."
+                "You are using the 'pyarrow' module, but the exact version is unknown "
+                "(possibly carried as an internal component by another module). Please "
+                f"make sure you are using pyarrow >= {MIN_PYARROW_VERSION}, < "
+                f"{MAX_PYARROW_VERSION} to ensure compatibility with Ray Datasets."
+                "If you want to disable this pyarrow version check, set the "
+                f"environment variable {RAY_DISABLE_PYARROW_VERSION_CHECK}=1."
             )
         else:
-            _VERSION_VALIDATED = True
+            version = pyarrow.__version__
+            if (
+                pkg_resources.packaging.version.parse(version)
+                < pkg_resources.packaging.version.parse(MIN_PYARROW_VERSION)
+            ) or (
+                pkg_resources.packaging.version.parse(version)
+                >= pkg_resources.packaging.version.parse(MAX_PYARROW_VERSION)
+            ):
+                raise ImportError(
+                    f"Datasets requires pyarrow >= {MIN_PYARROW_VERSION}, < "
+                    f"{MAX_PYARROW_VERSION}, but {version} is installed. Upgrade with "
```

Comment:
nit: maybe `Upgrade` -> `Reinstall`?

If user has a higher version like `pyarrow=9.0.0`, it is not an upgrade but downgrade.